graphqlapi/src/main/java/io/stargate/graphql/schema/cqlfirst/dml/fetchers/DmlFetcher.java

```diff
@@ -109,24 +110,64 @@ protected DmlFetcher(Table table, NameMapping nameMapping) {
     }
   }
 
-  protected List<Map<String, Object>> toListOfMutationResults(
-      ResultSet resultSet, List<Map<String, Object>> originalValues) {
-    List<Row> rows = resultSet.currentPageRows();
+  protected List<Map<String, Object>> toBatchResults(
+      List<Row> rows, List<Map<String, Object>> originalValues) {
+
+    boolean applied = isAppliedBatch(rows);
+
     List<Map<String, Object>> results = new ArrayList<>();
-    if (rows.isEmpty()) {
-      for (Map<String, Object> originalValue : originalValues) {
-        results.add(toAppliedMutationResultWithOriginalValue(originalValue));
-      }
-    } else {
-      for (int i = 0; i <= rows.size(); i++) {
-        Row row = rows.get(i);
-        Map<String, Object> originalValue = originalValues.get(i);
-        results.add(toMutationResultSingleRow(originalValue, row));
-      }
+    for (Map<String, Object> originalValue : originalValues) {
+      results.add(
+          applied
+              ? toAppliedMutationResultWithOriginalValue(originalValue)
+              : toUnappliedBatchResult(rows, originalValue));
     }
     return results;
   }
 
+  protected Map<String, Object> toBatchResult(List<Row> rows, Map<String, Object> originalValue) {
+    return isAppliedBatch(rows)
+        ? toAppliedMutationResultWithOriginalValue(originalValue)
+        : toUnappliedBatchResult(rows, originalValue);
+  }
+
+  private boolean isAppliedBatch(List<Row> batchRows) {
+    if (batchRows.isEmpty()) {
+      return true;
+    }
+    if (batchRows.size() > 1) {
+      return false;
+    }
+    Row row = batchRows.get(0);
+    return row.columns().stream().map(Column::name).anyMatch("[applied]"::equals)
+        && row.getBoolean("[applied]");
+  }
+
+  private Map<String, Object> toUnappliedBatchResult(
+      List<Row> rows, final Map<String, Object> originalValue) {
+    Map<String, Object> primaryKey =
+        table.primaryKeyColumns().stream()
+            .collect(
+                Collectors.toMap(
+                    Column::name, column -> toDBValue(column, originalValue.get(column.name()))));
+    return rows.stream()
+        .filter(row -> matches(row, primaryKey))
+        .findFirst()
```

Comment:
why only the first result is relevant?

Reply:
There can't be more than one result. A conditional batch response only contains each primary key once, even if there were multiple queries on that row (one of the integration tests covers that).